Rude/RudePlugin.cs

```diff
@@ -0,0 +1,119 @@
+
+namespace Rude
+{
+    public class RudePlugin : PluginBase
+    {
+        private readonly string[] Prefixes = { "[Rude] ", "[Ruderer] ", "[Rüdiger] " };
+        public const double MinRudeTime = 45;
+        public const double MaxRudeTime = 120;
+        public const double MinRudeActionTimeout = 3;
+        public const double MaxRudeActionTimeout = 18;
+
+        private ConcurrentDictionary<int, RudeStatus> RudeStatus = new ConcurrentDictionary<int, RudeStatus>();
+        private ConcurrentDictionary<int, RudeStatus> RudeActorStatus = new ConcurrentDictionary<int, RudeStatus>();
+
+        private Random rand = new Random();
+
+        [ContextCallback("Rude")]
+        public async Task RudeUser(IUser actor, IUserShim targetShim)
+        {
+            var target = await targetShim.Load();
+            EnsureRudeStatus(target.UserId);
+            EnsureRudeStatus(actor.UserId);
+
+            var rude = RudeStatus[target.UserId];
+            var rudeActor = RudeActorStatus[actor.UserId];
+
+            if (rudeActor.RudeLevel > 0)
+            {
+                var actorRude = RudeStatus[actor.UserId];
+
+                // You shouldn't be able to use this to reset yourself
+                if (rudeActor.RudeLevel < Prefixes.Length)
+                    actorRude.RaiseRudeLevel(RandomTimespan(MinRudeTime, MaxRudeTime));
+
+                return;
+            }
+
+            rudeActor.RaiseRudeLevel(RandomTimespan(MinRudeActionTimeout, MaxRudeActionTimeout));
+            rude.RaiseRudeLevel(RandomTimespan(MinRudeTime, MaxRudeTime));
+
+            if (rude.RudeLevel > Prefixes.Length)
+            {
+                rude.Reset();
+                await target.Kick("You have been too rude.");
+            }
+            else
+            {
+                target.Name = GetUserName(target);
+                await target.SaveChanges();
+            }
+        }
+
+        public override async Task OnUserConnected(IUser user)
+        {
+            user.Name = GetUserName(user);
+            await user.SaveChanges();
+        }
+
+        private void EnsureRudeStatus(int userId)
+        {
+            if (!RudeStatus.ContainsKey(userId))
+            {
+                RudeStatus.TryAdd(userId, new RudeStatus
+                {
+                    OnRudeLevelDecrease = async () =>
+                    {
+                        var users = await Server.GetOnlineUsers();
+                        var user = users.SingleOrDefault(a => a.UserId == userId);
+                        if (user == null)
+                            return;
+
+                        user.Name = GetUserName(user);
+                        await user.SaveChanges();
+                    }
+                });
+            }
+
+            if (!RudeActorStatus.ContainsKey(userId))
+            {
+                RudeActorStatus.TryAdd(userId, new RudeStatus
+                {
+                    OnRudeLevelDecrease = () => { }
```

Comment:
My opinion stands. Reusing this data structure for cooldowns feels **very** bolted-on so this callback still carries no value IMO. The class's one big feature is the delay mechanism where you can run code after a timeout and abstracting that *might* even make sense if you do it on a general layer (where it's no longer specific to rude) but that *still* doesn't give you a valid reason to use it *here* and without that, it's kinda pointless.